packages/cozy-realtime/src/CozyRealtime.js

```diff
@@ -1,360 +1,549 @@
 import MicroEE from 'microee'
-import Socket from './Socket'
-import minilog_ from 'minilog'
 
-const inBrowser = typeof window !== 'undefined'
-const minilog = (inBrowser && window.minilog) || minilog_
-const logger = minilog('cozy-realtime')
-minilog.suggest.deny('cozy-realtime', 'info')
-
-const INDEX_KEY_SEPARATOR = '//'
+import logger from './logger'
+import SubscriptionList from './SubscriptionList'
+import RetryManager from './RetryManager'
+import {
+  getUrl,
+  getToken,
+  doctype,
+  isInBrowser,
+  getCozyClientFromOptions,
+  isOnline
+} from './utils'
 
 /**
- * Generate a key for an event
- *
- * @return {String}  Event key
+ * Manage the realtime interactions with a cozy stack
  */
-export const generateKey = (eventName, type, id) =>
-  `${eventName}${INDEX_KEY_SEPARATOR}${type}${INDEX_KEY_SEPARATOR}${id}`
-
-const getHandlerAndId = (handlerOrId, handlerOrUndefined) => {
-  if (
-    typeof handlerOrId !== 'function' &&
-    typeof handlerOrUndefined !== 'function'
-  ) {
-    throw new Error('You should call this function with an handler')
+class CozyRealtime {
+  /**
+   * @constructor
+   * @param {object} options
+   * @param {CozyClient}  options.client
+   */
+  constructor(options) {
+    this.client = getCozyClientFromOptions(options)
+    this.subscriptions = new SubscriptionList()
+    this.retryManager = new RetryManager()
+    this.retryManager.on('error', err => this.emit('error', err))
+    this.bindEventHandlers()
   }
 
-  let id, handler
-  if (!handlerOrUndefined) {
-    id = undefined
-    handler = handlerOrId
-  } else {
-    id = handlerOrId
-    handler = handlerOrUndefined
+  /**
+   * Start all handlers and try to connect if online
+   * @private
+   */
+  start() {
+    if (!this.isStarted) {
+      logger.info('started')
+      this.isStarted = true
+      this.retryManager.reset()
+      this.subscribeGlobalEvents()
+      this.subscribeClientEvents()
+      this.emit('start')
+      if (isOnline()) this.connect()
+    }
   }
 
-  return { id, handler }
-}
-
-const VALID_EVENTS = ['created', 'updated', 'deleted', 'notified']
-
-const validateParameters = (eventName, type, id, handler) => {
-  let msg
-
-  if (!VALID_EVENTS.includes(eventName)) {
-    msg = `'${eventName}' is not a valid event, valid events are 'created', 'updated' or 'deleted'.`
+  /**
+   * Create and attach a WebSocket to the current instance
+   *
+   * @private
+   * @returns {WebSocket} the created websocket
+   */
+  createSocket() {
+    if (this.websocket) {
+      logger.warn(
+        'Unattended replacement of an existing socket, this should not happen'
+      )
+      this.revokeWebSocket(this.websocket)
```

Comment:
I think we can remove `this.websocket` from the arguments of this.removeWebSocket.

Reply:
:+1: